build-support/bin/generate_completions.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import json
+import subprocess
+
+BASH_COMPLETION_TEMPLATE = """
+# DO NOT EDIT.
+# This script is autogenerated by build-support/bin/generate_autocomplete.py
+
+readonly PANTS_GOALS="\\
+    {goals}"
+
+{goal_options}
+
+function pants_completions()
+{{
+    local current_word previous_word previous_goal
+
+    current_word=${{COMP_WORDS[COMP_CWORD]}}
+    previous_word=${{COMP_WORDS[COMP_CWORD-1]}}
+    previous_goal=$(get_previous_goal)
+
+    case "$previous_goal" in
+    {goal_cases}
+    esac
+
+    # Show goals, unless the tab completion has a hyphen - then show global/goal options instead
+    if [[ $current_word == -* ]]; then
+        COMPREPLY=( $( compgen -W "$complete_options" -- $current_word ))
+    else
+        COMPREPLY=( $( compgen -W "$PANTS_GOALS" -- $current_word ))
+    fi
+    return 0
+}}
+
+# Get the most recent goal in the command arguments, so options can be correctly applied
+# This function will ignore hyphenated options when looking for the goal
+# If this is empty, we're at the top-level Pants command
+function get_previous_goal()
+{{
+    local previous_goal i current_word
+    previous_goal=
+    for (( i=$COMP_CWORD; i > 0; --i )); do
+        current_word=${{COMP_WORDS[i]}}
+        if [[ $current_word != -* ]]; then
+            previous_goal=$current_word
+            break
+        fi
+    done
+    echo $previous_goal
+}}
+
+# TODO: Need to create a separate .zsh script instead of this
+autoload -U +X bashcompinit && bashcompinit
+autoload -Uz compinit && compinit
+
+complete -F pants_completions pants
+"""
+
+GOAL_OPTIONS_TEMPLATE = """
+readonly PANTS_{name}_OPTIONS="\\
+    {options}"
+"""
+
+GOAL_CASE_TEMPLATE = """
+    {goal})
+        complete_options="$PANTS_{bash_name}_OPTIONS"
+        ;;
+"""
+
+
+def _run_pants_help():
+    process = subprocess.run(["./pants", "help-all"], capture_output=True, check=True)
+    process.check_returncode()
+    return json.loads(process.stdout)
+
+
+def _parse_help_for_goal_names(help: dict) -> list[str]:
+    goals_dict = help["name_to_goal_info"]
+    return sorted(set(goals_dict.keys()))
+
+
+def _parse_help_for_goal_options(help: dict, goal_name: str) -> list[str]:
+    goal_help = help["scope_to_help_info"][goal_name]
+    # TODO: Scoped or unscoped? Scoped should probably be available globally as well
```

Comment:
Outstanding question. I'm currently using unscoped and it's tied to the previous goal. If I use scoped, I can technically run those from almost anywhere in the command line - which bloats the autocomplete. 

Even just all the global options really bloats auto-complete on tabbing, but that's probably acceptable for now.

Reply:
For the moment, unscoped options follow the goal, and global options, scoped goal options, and the `changed` subsystem's options are included in the global completion space